### ISSUE TITLE
Design a game module

### ISSUE THREAD

**Issue by borjasotomayor** [uchicago-cs/chiventure#20]
Design a module to model a single game, including functions to do basic operations on games (e.g., new/init, free, run a command, etc.)

**Comment by a-matai** [comment]
Attendees: @liuyifan-eric and @a-matai 

**Objectives:** 
Integrate new/init, free, run a command

**Summary:**
We've designed a simple skeleton for the game module right now with the following parts:
1. List of players (to support future design). Player structs contain all relevant information. One consideration here is that perhaps the list of objectives could be loaded into the game struct. As we are currently at one player, it does not matter where that information is right now?

2. Pointer to a struct that contains all the rooms (linked list with the node in this case being the current room. 
Given that a graph struct is overkill and most of the rooms will contain the information necessary, we just thought we would, in the game state, keep track of a list of all rooms. The order only matters insofar that the node is the room the player is currently in. We could eventually optimize so that the "most adjacent" are closer to the top. _We are looking into setting up a hashtable with the id so that traversal is faster as we are given room ids._

3. Time started (easy to tell time elapsed whenever needed later). This could, if functionality is warranted later, become a list of save times. 

The implementations of functions are pretty basic, with an init game that takes in a file (will check if there is a game or not, if not then will start new game, if yes then will load). Given we don't know how files will work yet this is skeleton code. Most functions themselves are skeletons, we will fill them up as we go along—the brunt of the effort has been thinking through precise design of the game struct.

**Comment by liuyifan-eric** [comment]
We updated the list of players and the list of rooms using hashtables provided in uthash.h. Now the the pointer to each player/room can be retrieved based on player/room id. 

Currently we removed the adjacency list design which saves the map of all the room, i.e. which rooms are connected to each room, because we would expect the room exit info (which room the player is going to when they goes through a door) is saved in the door object and moving to another room is by tracking the room id and using a hashtable. This feature may be reconsidered in the future based on the actual design of WDL team.

**Comment by liuyifan-eric** [comment]
Withdraw last comment. We finally decided to add adjacency list in each room struct, which stores a list of room_id's.

**Comment by liuyifan-eric** [comment]
The current structure of game_state looks like this:
   game
   |         \
room   player
   |         /
   object
Since players' inventory also contains objects, we decide to use the same object struct for room and player. 
We moved the hashtable of rooms (including the definitions and functions) to the room module. Similarly, in #21, hashtable of objects is in the object module, not the room module.

**Comment by a-matai** [comment]
**Closing comments**

The precise descriptions of the components of the game struct have been fleshed out and fulfilled, alongside some skeleton code to create a version that compiles (v0.0.0). Building out these skeletons into full functions is possible in the coming week as we will figure out how to initialize a game from WDL's information, and test out saving and loading with Checkpointing. One design issue that remains unresolved is the coordinates for rooms that would support a future map. Given the hashtable/llist design, there is a conflict between representations and dimensions for a possible future map. 

We ended up implementing a hashtable for the list of rooms (using uthash.h) instead of a linked list to make searching quicker. We also moved around struct definitions into the modules designed for #21 so that game.h includes the others, but the others don't include game.h. 


**Thoughts for the next sprint**
To-do from All-Hands Meeting Monday May 6 regd. game state 

- _WDL_: have built a parser, we need to create a function that takes rooms and creates hashtable.

- _Action Management:_ Decide functionalities and conditions for objects (they have their own struct, so we need to consolidate)

- _Checkpointing_ : Looked into two packages that would serialize and deserialize game-state, need to check-in and give them a working test game struct they can play with. 

- UI: Design game/room struct to be able to map rooms. Need to figure out how rooms are being stored, and if an adjacency list is enough information. Store images/etc is a game-state/WDL question
